cmd/snap-update-ns/utils.go

```diff
@@ -58,53 +141,31 @@ var (
 // after each segment is created and opened. The special value -1 may be used
 // to request that ownership is not changed.
 func secureMkdirAll(name string, perm os.FileMode, uid, gid int) error {
-	// Declare var and don't assign-declare below to ensure we don't swallow
-	// any errors by mistake.
-	var err error
-	var fd int
-
-	const openFlags = syscall.O_NOFOLLOW | syscall.O_CLOEXEC | syscall.O_DIRECTORY
+	logger.Debugf("secure-mkdir-all %q %v %d %d", name, perm, uid, gid)
 
 	// Only support absolute paths to avoid bugs in snap-confine when
 	// called from anywhere.
 	if !filepath.IsAbs(name) {
 		return fmt.Errorf("cannot create directory with relative path: %q", name)
 	}
-	// Open the root directory and start there.
-	fd, err = sysOpen("/", openFlags, 0)
+	segments := strings.FieldsFunc(filepath.Clean(name), func(c rune) bool { return c == '/' })
+
+	// Create the prefix.
+	fd, err := secureMkPrefix(segments, perm, uid, gid)
 	if err != nil {
-		return fmt.Errorf("cannot open root directory: %v", err)
+		return err
 	}
 	defer sysClose(fd)
 
-	// Split the path by entries and create each element using mkdirat() using
-	// the parent directory as reference. Each time we open the newly created
-	// segment using the O_NOFOLLOW and O_DIRECTORY flag so that symlink
-	// attacks are impossible to carry out.
-	segments := strings.FieldsFunc(filepath.Clean(name), func(c rune) bool { return c == '/' })
-	for _, segment := range segments {
-		made := true
-		if err = sysMkdirat(fd, segment, uint32(perm)); err != nil {
-			switch err {
-			case syscall.EEXIST:
-				made = false
-			default:
-				return fmt.Errorf("cannot mkdir path segment %q: %v", segment, err)
-			}
-		}
-		fd, err = sysOpenat(fd, segment, openFlags, 0)
+	if len(segments) > 0 {
+		// Create the final segment as a directory.
+		fd, err = secureMkDir(fd, segments, len(segments)-1, perm, uid, gid)
```

Comment:
Again, this is awkward without know the reason why you broke it up in this way. We do:
```
segments := ...
...
fd, err := secureMkPrefix(segments, perm, uid, gid)
...
fd, err = secureMkDir(fd, segments, len(segments)-1, perm, uid, gid)
```

To me, an API something along of this seems more clear:
```
// create the dir and any needed parent directories
createParents := true
fd, err := secureMkDir(segments, perm, uid, gid, createParents)
...
// create only the specified directory and no parents
createParents = false
fd, err := secureMkDir(segments, perm, uid, gid, createParents)
```

Then if you just want to create the parent directory (and its parents) for some reason, you just do:
```
fd, err := secureMkDir(segments[0:len(segments)-1], perm, uid, gid, true)
```